DemoInfo/DemoParser.cs

```diff
@@ -506,9 +527,142 @@ public void ParseHeader()
 
 			Header = header;
 
-
 			if (HeaderParsed != null)
 				HeaderParsed(this, new HeaderParsedEventArgs(Header));
+
+				if (header.PlaybackTime == 0)  //header is corrupted
+					FixHeader();
+		}
+
+		void FixHeader()
+		{
+			ExpressParseToEnd();
+			Header.PlaybackTicks = IngameTick;
+			Header.PlaybackFrames = CurrentTick - (TickAtStart + 1);
+			Header.PlaybackTime = TickInterval * IngameTick;
+			
+			CurrentTick = 0;
+
+			//the ol' switcheroo
+			//It's probably safest and easiest to
+			//swap out for a fresh bitstream.
+			Input.Seek(0, System.IO.SeekOrigin.Begin);
```

Comment:
Again, please correct me if I'm wrong here but AFAICS the only reason why you would need this is if *your code* needs these header fields from the start for some reason, right? Where is the ServerInfo message usually located in those demos? Because if it tends to be somewhere near the beginning, I still believe that most applications should be able to deal with this without restarting the parser.

---

I'm not sure if overwriting the DemoHeader fields is the best approach though. There should definitely be a new parser event that notifies you when we read the ServerInfo and I'd suggest to simply hand out the relevant data with this event.

It's a bit unfortunate that this shifts a part the burden to user code but as we saw, the only alternative is changing/breaking our API in a subtle way and I'd like to avoid that at all costs.